### ISSUE TITLE
(CA2022) How am I supposed to read correctly from a NetworkStream without knowing the incoming data size?

### ISSUE THREAD

**Issue by grallbring** [dotnet/roslyn-analyzers#7521]
Using this code snippet, I'm connecting to a TCP server, write a short message to it and then read it's response.
```cs
using (TcpClient tcpClient = new TcpClient())
{
    tcpClient.Connect("localhost", 8888);

    using (NetworkStream networkStream = tcpClient.GetStream())
    {
        networkStream.Write([0x01, 0x01]);

        byte[] data = new byte[1_024];

        networkStream.Read(data, 0, data.Length);
    }
}
```

Since I don't know the length of the response the server sends to me I create a buffer that I'm sure is large enough for it (1024 bytes).
Using `networkStream.Read(data, 0, data.Length);` to then read into that stream gives me a `CA2022` warning: `Avoid inexact read with 'System.Net.Sockets.NetworkStream.Read(byte[], int, int)'`, suggesting I should use `NetworkStream.ReadExact` instead.
This will block execution forever since the server never sends data that's 1024 bytes long to me.

My question is what the correct way is to do this without receiving this warning.
Thanks!

**Comment by mpidash** [comment]
You can check the return value of `networkStream.Read(data, 0, data.Length)` to see how much data you received. This is also suggested in the [doc](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2022#how-to-fix-violations) of CA2022 on how to fix a violation:

> To fix a violation, either check the return value (which is the total number of bytes read into the buffer) or ...

The fixer of CA2022 assumed that you always wanted to read exactly 1024 bytes (since you did not check the return value).
[Here](https://learn.microsoft.com/en-us/dotnet/core/compatibility/core-libraries/6.0/partial-byte-reads-in-streams#recommended-action) is another document that outlines recommended actions when working with `Stream.Read` and `Stream.ReadAsync`.
